From cc6d2c37b0237df92f8fb77fa15ae2dbda4ea989 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 3 Nov 2021 15:49:46 +0100
Subject: [PATCH 0630/1695] drivers: update counter framework to manage
 configuration

Add new interfaces to manage counter dynamic configuration.
It allows to change the counter input defined by the caller.
Aligned the lptimer driver accordingly.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2a45e255899d839aae91cd64258d1ffd8eda3799 drivers: update counter framework to manage configuration]
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Change-Id: Ib166047f3f83bb96d3969c55269a6621d9b9c31b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/223183
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/counter/counter.c       |  98 ++++++++++----
 core/drivers/counter/stm32_lptimer.c | 196 ++++++++++++++-------------
 core/include/drivers/counter.h       |  40 ++++--
 3 files changed, 205 insertions(+), 129 deletions(-)

diff --git a/core/drivers/counter/counter.c b/core/drivers/counter/counter.c
index 635aa87ba..a2c7b6ad5 100644
--- a/core/drivers/counter/counter.c
+++ b/core/drivers/counter/counter.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
- * Copyright (c) 2020, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2020-2021, STMicroelectronics - All Rights Reserved
  * Author(s): Ludovic Barre, <ludovic.barre@st.com> for STMicroelectronics.
  */
 
@@ -15,14 +15,25 @@ static unsigned int list_lock = SPINLOCK_UNLOCK;
 static LIST_HEAD(cnt_dev_list_head, counter_device) cnt_dev_list =
 	LIST_HEAD_INITIALIZER(cnt_dev_list_head);
 
-TEE_Result counter_start(struct counter_device *counter)
+TEE_Result counter_start(struct counter_device *counter, void *config)
 {
+	uint32_t exceptions = 0;
+
 	assert(counter && counter->ops);
 
 	if (!counter->ops->start)
 		return TEE_ERROR_NOT_SUPPORTED;
 
-	return counter->ops->start(counter);
+	exceptions = cpu_spin_lock_xsave(&list_lock);
+	if (counter->is_used) {
+		cpu_spin_unlock_xrestore(&list_lock, exceptions);
+		return TEE_ERROR_BUSY;
+	}
+
+	counter->is_used = true;
+	cpu_spin_unlock_xrestore(&list_lock, exceptions);
+
+	return counter->ops->start(counter, config);
 }
 
 /*
@@ -31,12 +42,22 @@ TEE_Result counter_start(struct counter_device *counter)
  */
 TEE_Result counter_stop(struct counter_device *counter)
 {
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t exceptions = 0;
+
 	assert(counter && counter->ops);
 
 	if (!counter->ops->stop)
 		return TEE_SUCCESS;
 
-	return counter->ops->stop(counter);
+	res = counter->ops->stop(counter);
+	if (!res) {
+		exceptions = cpu_spin_lock_xsave(&list_lock);
+		counter->is_used = false;
+		cpu_spin_unlock_xrestore(&list_lock, exceptions);
+	}
+
+	return res;
 }
 
 TEE_Result counter_get_value(struct counter_device *counter, unsigned int *ticks)
@@ -87,17 +108,27 @@ TEE_Result counter_cancel_alarm(struct counter_device *counter)
 	return res;
 }
 
+void counter_release_config(struct counter_device *counter,
+			    void *config)
+{
+	assert(counter && config);
+
+	if (counter->ops->release_config)
+		counter->ops->release_config(config);
+}
+
 #ifdef CFG_DT
-struct counter_device *fdt_counter_get(const void *fdt, int offs)
+struct counter_device *fdt_counter_get(const void *fdt, int offs, void **config)
 {
 	struct counter_device *cnt_dev = NULL;
-	const fdt32_t *cuint;
-	uint32_t exceptions;
+	const fdt32_t *cuint = NULL;
+	int len = 0;
+	uint32_t exceptions = 0;
 	int phandle = 0;
 
-	assert(fdt);
+	assert(fdt && config);
 
-	cuint = fdt_getprop(fdt, offs, "counter", NULL);
+	cuint = fdt_getprop(fdt, offs, "counter", &len);
 	if (!cuint)
 		return NULL;
 
@@ -112,15 +143,40 @@ struct counter_device *fdt_counter_get(const void *fdt, int offs)
 	if (!cnt_dev)
 		EMSG("No counter device registered for phandle: %d", phandle);
 
-	if (cnt_dev && !cnt_dev->is_used)
-		cnt_dev->is_used = true;
-	else
-		cnt_dev = NULL;
+	if ((uint32_t)len < sizeof(uint32_t))
+		goto out;
+
+	len -= sizeof(uint32_t);
+	cuint++;
+
+	/* Len in bytes */
+	len /= sizeof(uint32_t);
+
+	if (cnt_dev && cnt_dev->ops->set_config)
+		cnt_dev->ops->set_config(cnt_dev, cuint, len, config);
 
+out:
 	cpu_spin_unlock_xrestore(&list_lock, exceptions);
 
 	return cnt_dev;
 }
+#else
+TEE_Result counter_get_config(struct counter_device *cnt_dev,
+			      const struct counter_param *param,
+			      void **config)
+{
+	if (!param)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!cnt_dev->ops->set_config)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (cnt_dev && cnt_dev->ops->set_config)
+		cnt_dev->ops->set_config(cnt_dev, param->params,
+					 param->len, config);
+
+	return TEE_SUCCESS;
+}
 #endif
 
 struct counter_device *counter_get_by_name(const char *name)
@@ -137,27 +193,11 @@ struct counter_device *counter_get_by_name(const char *name)
 			break;
 	}
 
-	if (cnt_dev && !cnt_dev->is_used)
-		cnt_dev->is_used = true;
-	else
-		cnt_dev = NULL;
-
 	cpu_spin_unlock_xrestore(&list_lock, exceptions);
 
 	return cnt_dev;
 }
 
-void counter_put(struct counter_device *counter)
-{
-	uint32_t exceptions;
-
-	assert(counter);
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	counter->is_used = false;
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-}
-
 struct counter_device *counter_dev_alloc(void)
 {
 	return calloc(1, sizeof(struct counter_device));
diff --git a/core/drivers/counter/stm32_lptimer.c b/core/drivers/counter/stm32_lptimer.c
index f12a95d4f..b8df21920 100644
--- a/core/drivers/counter/stm32_lptimer.c
+++ b/core/drivers/counter/stm32_lptimer.c
@@ -124,6 +124,11 @@ struct lptimer_device {
 	struct itr_handler *itr;
 };
 
+struct stm32_lptimer_config {
+	struct lptimer_ext_input_cfg ext_input;
+	struct lptimer_ext_trigger_cfg ext_trigger;
+};
+
 static TEE_Result stm32_lpt_counter_set_alarm(struct counter_device *counter)
 {
 	struct lptimer_device *lpt_dev = counter_priv(counter);
@@ -270,31 +275,33 @@ static TEE_Result stm32_lpt_counter_stop(struct counter_device *counter)
 	return TEE_SUCCESS;
 }
 
-static TEE_Result stm32_lpt_counter_start(struct counter_device *counter)
+static TEE_Result stm32_lpt_counter_start(struct counter_device *counter,
+					  void *config)
 {
 	struct lptimer_device *lpt_dev = counter_priv(counter);
 	struct stm32_lptimer_platdata *pdata = &lpt_dev->pdata;
-	struct lptimer_ext_input_cfg *ext_input = &pdata->ext_input;
-	struct lptimer_ext_trigger_cfg *ext_trigger = &pdata->ext_trigger;
+	struct stm32_lptimer_config *conf =
+		(struct stm32_lptimer_config *)config;
 	uintptr_t base = pdata->base;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	uint32_t cfgr = 0;
 	uint32_t cfgr2 = 0;
 
-	if (ext_input->num) {
+	if (conf && conf->ext_input.num) {
 		cfgr += _LPTIM_CFGR_COUNTMODE;
 
 		cfgr2 += _LPTIM_FLD_PREP(_LPTIM_CFGR2_IN1SEL,
-					 ext_input->mux[0]);
-		if (ext_input->num == 2)
+					 conf->ext_input.mux[0]);
+		if (conf->ext_input.num == 2)
 			cfgr2 += _LPTIM_FLD_PREP(_LPTIM_CFGR2_IN2SEL,
-						 ext_input->mux[1]);
+						 conf->ext_input.mux[1]);
 	}
 
-	if (ext_trigger->num) {
-		cfgr += _LPTIM_FLD_PREP(_LPTIM_CFGR_TRIGSEL, ext_trigger->mux);
+	if (conf && conf->ext_trigger.num) {
+		cfgr += _LPTIM_FLD_PREP(_LPTIM_CFGR_TRIGSEL,
+					conf->ext_trigger.mux);
 		cfgr += _LPTIM_FLD_PREP(_LPTIM_CFGR_TRIGEN,
-					ext_trigger->polarity);
+					conf->ext_trigger.polarity);
 		cfgr |= _LPTIM_CFGR_TIMOUT;
 	}
 
@@ -316,12 +323,99 @@ static TEE_Result stm32_lpt_counter_start(struct counter_device *counter)
 	return TEE_SUCCESS;
 }
 
+#ifdef CFG_EMBED_DTB
+static uint32_t get_param(const void *params, unsigned int step)
+{
+	const fdt32_t *param_dt = (fdt32_t *)params;
+
+	assert(param_dt);
+
+	return fdt32_to_cpu(*(param_dt + step));
+}
+#else
+static uint32_t get_param(const void *params, unsigned int step)
+{
+	const uint32_t *tim_param = (uint32_t *)params;
+
+	assert(tim_param);
+
+	return tim_params[step];
+}
+#endif
+
+static TEE_Result
+stm32_lptimer_counter_set_config(struct counter_device *counter,
+				 const void *param,
+				 int len, void **config)
+{
+	struct lptimer_device *lpt_dev = counter_priv(counter);
+	struct lptimer_driver_data *ddata = lpt_dev->ddata;
+	struct stm32_lptimer_config *conf;
+	struct lptimer_ext_input_cfg *ext_input;
+	struct lptimer_ext_trigger_cfg *ext_trigger;
+
+	if (!param)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (len < 4 || len > 5)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	conf = calloc(1, sizeof(*conf));
+	if (!conf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	ext_trigger = &conf->ext_trigger;
+	ext_trigger->polarity = get_param(param, 0);
+	ext_trigger->mux = get_param(param, 1);
+	ext_trigger->num = 1;
+
+	if (ext_trigger->mux >= ddata->nb_ext_trigger) {
+		free(conf);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	DMSG("external trigger pol:%"PRId8" mux:%"PRId8,
+	     ext_trigger->polarity, ext_trigger->mux);
+
+	/* external-input = < polarity mux1 mux2 > */
+	ext_input = &conf->ext_input;
+	ext_input->num = len - 3;
+	ext_input->polarity = get_param(param, 2);
+	ext_input->mux[0] = get_param(param, 3);
+
+	if (ext_input->num == 2)
+		ext_input->mux[1] = get_param(param, 4);
+
+	if (ext_input->mux[0] >= ddata->nb_ext_input ||
+	    ext_input->mux[1] >= ddata->nb_ext_input) {
+		free(conf);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	DMSG("external input nb:%"PRIu8" pol:%"PRIu8" mux[0]:%"PRIu8" mux[1]:%"PRIu8,
+	     ext_input->num, ext_input->polarity,
+	     ext_input->mux[0], ext_input->mux[1]);
+
+	*config = conf;
+
+	return TEE_SUCCESS;
+}
+
+static void stm32_lptimer_counter_release_config(void *config)
+{
+	struct lptimer_config *conf = (struct lptimer_config *)config;
+
+	free(conf);
+}
+
 static const struct counter_ops stm32_lpt_counter_ops = {
 	.start = stm32_lpt_counter_start,
 	.stop = stm32_lpt_counter_stop,
 	.get_value = stm32_lpt_counter_get_value,
 	.set_alarm = stm32_lpt_counter_set_alarm,
 	.cancel_alarm = stm32_lpt_counter_cancel_alarm,
+	.set_config = stm32_lptimer_counter_set_config,
+	.release_config = stm32_lptimer_counter_release_config,
 };
 
 static enum itr_return stm32_lptimer_itr(struct itr_handler *h)
@@ -392,78 +486,6 @@ struct dt_id_attr {
 	fdt32_t id_attr[1];
 };
 
-/* external-input = < polarity mux1 mux2 > */
-static int stm32_lptimer_ext_input_dt(struct lptimer_device *lpt_dev)
-{
-	struct lptimer_ext_input_cfg *ext_input = &lpt_dev->pdata.ext_input;
-	struct lptimer_driver_data *ddata = lpt_dev->ddata;
-	const uint32_t *cuint = NULL;
-	int len = 0;
-
-	cuint = fdt_getprop(lpt_dev->fdt, lpt_dev->node,
-			    "external-input", &len);
-	ext_input->num = 0;
-	ext_input->mux[0] = 0;
-	ext_input->mux[1] = 0;
-
-	if (!cuint)
-		return 0;
-
-	len /= sizeof(uint32_t);
-	if (len < 2 || len > 3)
-		return -LPTIM_ERR_INVAL;
-
-	ext_input->num = len - 1;
-	ext_input->polarity = fdt32_to_cpu(*cuint++);
-	ext_input->mux[0] = fdt32_to_cpu(*cuint++);
-
-	if (ext_input->num == 2)
-		ext_input->mux[1] = fdt32_to_cpu(*cuint);
-
-	if (ext_input->mux[0] >= ddata->nb_ext_input ||
-	    ext_input->mux[1] >= ddata->nb_ext_input)
-		return -LPTIM_ERR_INVAL;
-
-	DMSG("external input nb:%"PRIu8" pol:%"PRIu8" mux[0]:%"PRIu8" mux[1]:%"PRIu8,
-	     ext_input->num, ext_input->polarity,
-	     ext_input->mux[0], ext_input->mux[1]);
-
-	return 0;
-}
-
-/* external-trigger = < polarity mux > */
-static int stm32_lptimer_ext_trigger_dt(struct lptimer_device *lpt_dev)
-{
-	struct stm32_lptimer_platdata *pdata = &lpt_dev->pdata;
-	struct lptimer_driver_data *ddata = lpt_dev->ddata;
-	struct lptimer_ext_trigger_cfg *ext_trigger = &pdata->ext_trigger;
-	const uint32_t *cuint = NULL;
-	int len = 0;
-
-	cuint = fdt_getprop(lpt_dev->fdt, lpt_dev->node,
-			    "external-trigger", &len);
-	ext_trigger->num = 0;
-
-	if (!cuint)
-		return 0;
-
-	len /= sizeof(uint32_t);
-	if (len != 2)
-		return -LPTIM_ERR_INVAL;
-
-	ext_trigger->polarity = fdt32_to_cpu(*cuint++);
-	ext_trigger->mux = fdt32_to_cpu(*cuint);
-	ext_trigger->num = 1;
-
-	if (ext_trigger->mux >= ddata->nb_ext_trigger)
-		return -LPTIM_ERR_INVAL;
-
-	DMSG("external trigger pol:%"PRId8" mux:%"PRId8,
-	     ext_trigger->polarity, ext_trigger->mux);
-
-	return 0;
-}
-
 static TEE_Result stm32_lptimer_parse_fdt(struct lptimer_device *lpt_dev)
 {
 	struct stm32_lptimer_platdata *pdata = &lpt_dev->pdata;
@@ -473,7 +495,6 @@ static TEE_Result stm32_lptimer_parse_fdt(struct lptimer_device *lpt_dev)
 	int node = lpt_dev->node;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	int len = 0;
-	int err = 0;
 
 	assert(lpt_dev && lpt_dev->fdt);
 
@@ -498,14 +519,6 @@ static TEE_Result stm32_lptimer_parse_fdt(struct lptimer_device *lpt_dev)
 	if (res)
 		return res;
 
-	err = stm32_lptimer_ext_trigger_dt(lpt_dev);
-	if (err)
-		return err;
-
-	err = stm32_lptimer_ext_input_dt(lpt_dev);
-	if (err)
-		return err;
-
 	node = fdt_subnode_offset(fdt, node, "counter");
 	if (node >= 0 &&
 	    !fdt_node_check_compatible(fdt, node, LPTIMER_COUNTER_COMPAT) &&
@@ -637,3 +650,4 @@ DEFINE_DT_DRIVER(stm32_lptimer_dt_driver) = {
 	.match_table = stm32_lptimer_match_table,
 	.probe = stm32_lptimer_probe,
 };
+
diff --git a/core/include/drivers/counter.h b/core/include/drivers/counter.h
index 8f625cb38..3689756ff 100644
--- a/core/include/drivers/counter.h
+++ b/core/include/drivers/counter.h
@@ -1,11 +1,12 @@
 /* SPDX-License-Identifier: BSD-2-Clause */
 /*
- * Copyright (c) 2020, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2020-2021, STMicroelectronics - All Rights Reserved
  */
 
 #ifndef COUNTER_H
 #define COUNTER_H
 
+#include <libfdt.h>
 #include <stdint.h>
 #include <stdbool.h>
 #include <sys/queue.h>
@@ -28,12 +29,21 @@ struct alarm_cfg {
 
 struct counter_device;
 
+struct counter_param {
+	uint32_t *params;
+	size_t len;
+};
+
 struct counter_ops {
-	TEE_Result (*start)(struct counter_device *counter);
+	TEE_Result (*start)(struct counter_device *counter, void *config);
 	TEE_Result (*stop)(struct counter_device *counter);
 	TEE_Result (*get_value)(struct counter_device *counter, unsigned int *ticks);
 	TEE_Result (*set_alarm)(struct counter_device *counter);
 	TEE_Result (*cancel_alarm)(struct counter_device *counter);
+	TEE_Result (*set_config)(struct counter_device *counter,
+				 const void *param,
+				 int len, void **config);
+	void (*release_config)(void *config);
 };
 
 /**
@@ -62,7 +72,7 @@ struct counter_device {
 /**
  * @brief Start counter device in free running mode.
  */
-TEE_Result counter_start(struct counter_device *counter);
+TEE_Result counter_start(struct counter_device *counter, void *config);
 
 /**
  * @brief Stop counter device.
@@ -84,13 +94,30 @@ TEE_Result counter_set_alarm(struct counter_device *counter);
  */
 TEE_Result counter_cancel_alarm(struct counter_device *counter);
 
+/**
+ * @brief Release the counter configuration.
+ */
+void counter_release_config(struct counter_device *counter, void *config);
+
 #ifdef CFG_DT
 /**
  * @brief Parse and lookup a counter referenced by a device node.
+ * Retrieve an associated configuration.
  *
  * @retval counter device if successful, else 0 on error.
  */
-struct counter_device *fdt_counter_get(const void *fdt, int offs);
+struct counter_device *fdt_counter_get(const void *fdt,
+				       int offs, void **config);
+#else
+/**
+ * @brief Give the counter associated configuration link to given
+ * parameters.
+ *
+ * @retval TEE_SUCCESS if config is returned, error value otherwise.
+ */
+TEE_Result counter_get_config(struct counter_device *cnt_dev,
+			      const struct counter_param *param,
+			      void **config);
 #endif
 
 /**
@@ -100,11 +127,6 @@ struct counter_device *fdt_counter_get(const void *fdt, int offs);
  */
 struct counter_device *counter_get_by_name(const char *name);
 
-/**
- * @brief consumer release the counter device.
- */
-void counter_put(struct counter_device *counter);
-
 /* API for provider */
 static inline void *counter_priv(struct counter_device *counter)
 {
-- 
2.47.2

